OrbitSsh/Session.cpp

```diff
@@ -9,138 +9,111 @@
 }
 
-Session::~Session() {
-  if (raw_session_ptr_ != nullptr) {
-    libssh2_session_free(raw_session_ptr_);
+outcome::result<void> Session::Handshake(Socket* socket_ptr) {
+  const int rc = libssh2_session_handshake(raw_session_ptr_.get(),
+                                           socket_ptr->GetFileDescriptor());
+  if (rc == 0) {
+    return outcome::success();
+  } else {
+    return static_cast<Error>(rc);
   }
 }
 
-ResultType Session::Handshake(Socket* socket_ptr) {
-  int rc = libssh2_session_handshake(raw_session_ptr_,
-                                     socket_ptr->GetFileDescriptor());
-  if (rc == 0) return ResultType::kSuccess;
-  if (rc == LIBSSH2_ERROR_EAGAIN) return ResultType::kAgain;
-
-  ERROR("Unable to perform session handshake, error code %d", rc);
-  return ResultType::kError;
-}
-
-ResultType Session::MatchKnownHosts(std::string host, int port,
-                                    std::filesystem::path known_hosts_path) {
-  LIBSSH2_KNOWNHOSTS* known_hosts = libssh2_knownhost_init(raw_session_ptr_);
+outcome::result<void> Session::MatchKnownHosts(
+    std::string host, int port, std::filesystem::path known_hosts_path) {
+  LIBSSH2_KNOWNHOSTS* known_hosts =
+      libssh2_knownhost_init(raw_session_ptr_.get());
 
   if (known_hosts == nullptr) {
-    int last_error_code = libssh2_session_last_errno(raw_session_ptr_);
-    ERROR(
-        "Unable to initialize ssh known hosts object, last error code "
-        "%d",
-        last_error_code);
-    return ResultType::kError;
+    return static_cast<Error>(
+        libssh2_session_last_errno(raw_session_ptr_.get()));
   }
 
   // libssh2 does not support anything else than:
   // LIBSSH2_KNOWNHOST_FILE_OPENSSH
-  int amount_hosts =
+  const int amount_hosts =
       libssh2_knownhost_readfile(known_hosts, known_hosts_path.string().c_str(),
                                  LIBSSH2_KNOWNHOST_FILE_OPENSSH);
   if (amount_hosts < 0) {
-    ERROR("Unable to read known hosts file, error code: %d, file path %s",
-          amount_hosts, known_hosts_path.c_str());
     libssh2_knownhost_free(known_hosts);
-    return ResultType::kError;
+    return static_cast<Error>(amount_hosts);
   }
 
   size_t fingerprint_length;
   int fingerprint_type;
   const char* fingerprint = libssh2_session_hostkey(
-      raw_session_ptr_, &fingerprint_length, &fingerprint_type);
+      raw_session_ptr_.get(), &fingerprint_length, &fingerprint_type);
+
   if (fingerprint == nullptr) {
-    int last_error_code = libssh2_session_last_errno(raw_session_ptr_);
-    ERROR("Unable to get host fingerprint, last error code: %d",
-          last_error_code);
     libssh2_knownhost_free(known_hosts);
-    return ResultType::kError;
+    return static_cast<Error>(
+        libssh2_session_last_errno(raw_session_ptr_.get()));
   }
 
-  struct libssh2_knownhost* known_host;
-  int check_result = libssh2_knownhost_checkp(
+  const int check_result = libssh2_knownhost_checkp(
```

Comment:
this can return one of these errors. (Not part of Error.h)
LIBSSH2_KNOWNHOST_CHECK_FAILURE
LIBSSH2_KNOWNHOST_CHECK_NOTFOUND
LIBSSH2_KNOWNHOST_CHECK_MATCH
LIBSSH2_KNOWNHOST_CHECK_MISMATCH

Reply:
Thanks for spotting this. I added a new error category `KnowHostsError`